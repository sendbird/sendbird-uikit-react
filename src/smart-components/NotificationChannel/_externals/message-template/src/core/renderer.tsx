import React from 'react';

import type { BasicProps, ComponentType, ComponentsUnion, GetProperties } from '../types/components';

type FC<Props, ParsedStyle> = (props: BasicProps<Props, ParsedStyle>) => React.ReactElement | null;

type ComponentProperties<Type extends ComponentType> = GetProperties<Type, ComponentsUnion>;

interface RendererOptions<ParsedStyle> {
  views?: {
    [component in ComponentType]?: FC<ComponentProperties<component>, ParsedStyle>;
  };
}

export type Renderer<ParsedStyle> = {
  [component in ComponentType]: FC<ComponentProperties<component>, ParsedStyle>;
};

const FRAGMENT = ({ children }: React.PropsWithChildren) => <React.Fragment>{children}</React.Fragment>;

export function createRenderer<ParsedStyle = object | string>(
  opts?: RendererOptions<ParsedStyle>,
): Renderer<ParsedStyle> {
  return {
    box: opts?.views?.box || FRAGMENT,
    text: opts?.views?.text || FRAGMENT,
    image: opts?.views?.image || FRAGMENT,
    imageButton: opts?.views?.imageButton || FRAGMENT,
    textButton: opts?.views?.textButton || FRAGMENT,
  };
}
