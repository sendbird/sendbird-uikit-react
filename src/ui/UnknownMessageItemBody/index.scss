@import '../../styles/variables';

.sendbird-unknown-message-item-body {
  position: relative;
  display: inline-block;
  box-sizing: border-box;
  padding: 8px 12px;
  border-radius: 16px;
  &.reactions {
    border-radius: 16px 16px 0px 0px;
  }
  &.outgoing {
    @include themed() { background-color: t(primary-3) }
  }
  &.incoming {
    @include themed() { background-color: t(bg-1) }
  }
  &.mouse-hover, &:hover {
    &.outgoing {
      @include themed() { background-color: t(primary-4) }
    }
    &.incoming {
      @include themed() { background-color: t(bg-2) }
    }
  }

}

.sendbird-unknown-message-item-body__header {
  display: block;
}

.sendbird-unknown-message-item-body__description {
  display: block;
}
