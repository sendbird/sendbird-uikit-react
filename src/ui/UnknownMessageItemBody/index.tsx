import React, { ReactElement, useContext } from 'react';
import './index.scss';

import Label, { LabelTypography, LabelColors } from '../Label';
import { getClassName } from '../../utils';
import { LocalizationContext } from '../../lib/LocalizationContext';
import { BaseMessage } from '@sendbird/chat/message';

interface Props {
  className?: string | Array<string>;
  isByMe?: boolean;
  message: BaseMessage;
  customText?: string;
  mouseHover?: boolean;
  isReactionEnabled?: boolean;
}

export default function UnknownMessageItemBody({
  className,
  message,
  customText,
  isByMe = false,
  mouseHover = false,
  isReactionEnabled = false,
}: Props): ReactElement {
  const { stringSet } = useContext(LocalizationContext);
  return (
    <div className={getClassName([
      className,
      'sendbird-unknown-message-item-body',
      isByMe ? 'outgoing' : 'incoming',
      mouseHover ? 'mouse-hover' : '',
      (isReactionEnabled && message?.reactions?.length > 0) ? 'reactions' : '',
    ])}>
      {
        !customText && (
          <Label
            className="sendbird-unknown-message-item-body__header"
            type={LabelTypography.BODY_1}
            color={isByMe ? LabelColors.ONCONTENT_1 : LabelColors.ONBACKGROUND_1}
          >
            {stringSet.UNKNOWN__UNKNOWN_MESSAGE_TYPE}
          </Label>
        )
      }
      <Label
        className="sendbird-unknown-message-item-body__description"
        type={LabelTypography.BODY_1}
        color={isByMe ? LabelColors.ONCONTENT_2 : LabelColors.ONBACKGROUND_2}
      >
        {customText || stringSet.UNKNOWN__CANNOT_READ_MESSAGE}
      </Label>
    </div>
  );
}
