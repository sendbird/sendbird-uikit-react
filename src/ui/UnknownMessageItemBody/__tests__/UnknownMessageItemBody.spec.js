import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import UnknownMessageItemBody from "../index";

const createMockMessage = (process) => {
  const mockMessage = {
    reactions: [],
  };
  return process ? process(mockMessage) : mockMessage;
};

describe('UnknownMessageItemBody', () => {
  it('should have class names of own basic status', () => {
    const className = 'class-name-for-test';
    const component = mount(
      <UnknownMessageItemBody
        className={className}
        message={createMockMessage()}
      />
    );
    expect(
      component.find('.sendbird-unknown-message-item-body').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-unknown-message-item-body__header').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-unknown-message-item-body__description').hostNodes().exists()
    ).toBe(true);

    expect(
      component.find(`.${className}.sendbird-unknown-message-item-body`).hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-unknown-message-item-body.incoming').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-unknown-message-item-body.outgoing').hostNodes().exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-unknown-message-item-body.mouse-hover').hostNodes().exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-unknown-message-item-body.reactions').hostNodes().exists()
    ).toBe(false);
  });

  it('should have class name by isByMe', () => {
    const outgoingMessage = mount(
      <UnknownMessageItemBody
        message={createMockMessage()}
        isByMe
      />
    );
    expect(
      outgoingMessage.find('.sendbird-unknown-message-item-body.outgoing').hostNodes().exists()
    ).toBe(true);
    expect(
      outgoingMessage.find('.sendbird-unknown-message-item-body.incoming').hostNodes().exists()
    ).toBe(false);
    const incomingMessage = mount(
      <UnknownMessageItemBody
        message={createMockMessage()}
        isByMe={false}
      />
    );
    expect(
      incomingMessage.find('.sendbird-unknown-message-item-body.outgoing').hostNodes().exists()
    ).toBe(false);
    expect(
      incomingMessage.find('.sendbird-unknown-message-item-body.incoming').hostNodes().exists()
    ).toBe(true);
  });

  it('should have class name by mouseHover prop', () => {
    const component = mount(
      <UnknownMessageItemBody
        message={createMockMessage()}
        mouseHover
      />
    );
    expect(
      component.find('.sendbird-unknown-message-item-body.mouse-hover').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-unknown-message-item-body.incoming.mouse-hover').hostNodes().exists()
    ).toBe(true);
  });

  it('should have class name by reactions of message', () => {
    const component = mount(
      <UnknownMessageItemBody
        isReactionEnabled
        message={createMockMessage((mockMessage) => ({
          ...mockMessage,
          reactions: [{}, {}, {}],
        }))}
      />
    );
    expect(
      component.find('.sendbird-unknown-message-item-body.reactions').hostNodes().exists()
    ).toBe(true);
  });

  it('should do a snapshot test of the UnknownMessageItemBody DOM', function () {
    const component = renderer.create(
      <UnknownMessageItemBody
        className="class-name-for-snapshot"
        message={createMockMessage((mock) => ({
          ...mock,
          reactions: [{}, {}, {}],
        }))}
        isByMe
        mouseHover
      />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
