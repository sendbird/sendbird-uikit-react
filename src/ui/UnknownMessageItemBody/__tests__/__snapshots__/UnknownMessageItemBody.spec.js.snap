// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnknownMessageItemBody should do a snapshot test of the UnknownMessageItemBody DOM 1`] = `
<div
  className="class-name-for-snapshot sendbird-unknown-message-item-body outgoing mouse-hover "
>
  <span
    className="sendbird-unknown-message-item-body__header sendbird-label sendbird-label--body-1 sendbird-label--color-oncontent-1"
  >
    (Unknown message type)
  </span>
  <span
    className="sendbird-unknown-message-item-body__description sendbird-label sendbird-label--body-1 sendbird-label--color-oncontent-2"
  >
    Cannot read this message.
  </span>
</div>
`;
